redash/models.py

```diff
@@ -1034,7 +1012,134 @@ def record(cls, event):
+                configuration[prop] = current_configuration[prop]
+
+    def add_group(self, group, view_only=False):
+        ndg = NotificationDestinationGroup.create(group=group, notification_destination=self, view_only=view_only)
+        setattr(self, 'notification_destination_groups', ndg)
+
+    def remove_group(self, group):
+        NotificationDestinationGroup.delete().where(NotificationDestinationGroup.group==group, NotificationDestinationGroup.notification_destination==self).execute()
+
+    def update_group_permission(self, group, view_only):
+        ndg = NotificationDestinationGroup.get(NotificationDestinationGroup.group==group, NotificationDestinationGroup.notification_destination==self)
+        ndg.view_only = view_only
+        ndg.save()
+        setattr(self, 'notification_destination_groups', ndg)
+
+    @property
+    def destination(self):
+        return get_destination(self.type, self.options)
+
+    @classmethod
+    def all(cls, org, groups=None):
+        notification_destinations = cls.select().where(cls.org==org).order_by(cls.id.asc())
+
+        if groups:
+            notification_destinations = notification_destinations.join(NotificationDestinationGroup).where(NotificationDestinationGroup.group << groups)
+
+        return notification_destinations
+
+    @property
+    def groups(self):
+        groups = NotificationDestinationGroup.select().where(NotificationDestinationGroup.notification_destination==self)
+        return dict(map(lambda g: (g.group_id, g.view_only), groups))
+
+
+class NotificationDestinationGroup(BaseModel):
+    notification_destination = peewee.ForeignKeyField(NotificationDestination)
+    group = peewee.ForeignKeyField(Group, related_name="notification_destinations")
+    view_only = peewee.BooleanField(default=False)
+
+    class Meta:
+        db_table = "notification_destination_groups"
+
+
+class AlertSubscription(ModelTimestampsMixin, BaseModel):
+    user = peewee.ForeignKeyField(User)
+    destination = peewee.ForeignKeyField(NotificationDestination, null=True)
+    alert = peewee.ForeignKeyField(Alert)
+
+    class Meta:
+        db_table = 'alert_subscriptions'
+
+    def to_dict(self):
+        return {
+            'user': self.user.to_dict(),
+            'destination': self.destination.to_dict(),
+            'alert_id': self.alert_id
+        }
+
+    @classmethod
+    def all(cls, alert_id):
+        return AlertSubscription.select(AlertSubscription, User).join(User).where(AlertSubscription.alert==alert_id)
+
+    @classmethod
+    def unsubscribe(cls, alert_id, user_id):
+        query = AlertSubscription.delete().where(AlertSubscription.alert==alert_id).where(AlertSubscription.user==user_id)
+        return query.execute()
+
+
+all_models = (Organization, Group, DataSource, DataSourceGroup, User, QueryResult, Query, Alert, AlertSubscription, Dashboard, Visualization, Widget, Event, NotificationDestination, NotificationDestinationGroup)
```

Comment:
Because of the foreign keys relationships the order here is important, and `NotificationDestination` and `NotificationDestinationGroup` should be before `AlertSubscription`.